apps/vmq_server/src/vmq_reg.erl

```diff
@@ -492,32 +501,90 @@ fix_dead_queue(SubscriberId, Subs, {DeadNodes, [Target|Targets], N}) ->
     %%%  and ensure that a queue exist for such subscriptions.
     %%%  In case allow_multiple_sessions=false (default) all
     %%%  subscriptions will be remapped
-    NewSubs =
+
+    %% check if there's any work to do:
+    NewSubs = rewrite_dead_nodes(Subs, DeadNodes, '$notanodename', false),
+    case Subs of
+        NewSubs ->
+            %% no subs were changed, nothing to do here.
+            {DeadNodes, [Target|Targets], N};
+        _ ->
+            %% Assume all the subscriptions come with the same clean_session flag.
+            [{_, StartClean, _}|_] = Subs,
+            RepairQueueFun =
+                fun() ->
+                        case rpc:call(Target, vmq_reg, replace_dead_queue, [SubscriberId, DeadNodes, StartClean, Subs]) of
+                            ok ->
+                                {DeadNodes, Targets ++ [Target], N + 1};
+                            Error ->
+                                lager:format("repairing dead queue for ~p on ~p failed due to ~p~n", [SubscriberId, Target, Error]),
+                                {DeadNodes, Targets ++ [Target], N}
+                        end
+                end,
+            vmq_reg_sync:sync(SubscriberId, RepairQueueFun, 60000)
+    end.
+
+
+replace_dead_queue(SubscriberId, _DeadNodes, _StartClean = true, _Subs) ->
+    %% At this point we may have a new subscriber on either this or
+    %% other nodes which may have another set of subscriptions than
+    %% what the old node had.
+
+    %% To ensure we end up with a client which has a consistent state,
+    %% we write an empty subscription to the store which will cause
+    %% clients on other nodes to be disconnected (if
+    %% allow_multiple_sessions=false) (TODO: formalize this behaviour
+    %% in a test-case) and they'll then reconnect.
+    Subs = vmq_subscriber:new(true),
+    vmq_subscriber_db:store(SubscriberId, Subs),
```

Comment:
aren't we create another race with the `get_queue_pid` below?

Reply:
I think you're right, it would probably be better to put the `store(SId, Subs)`  into the each of the two branches of the case statement.